platform_code/stm32cube/serial/micro_ros_transport.cpp

```diff
@@ -0,0 +1,104 @@
+#include <micro_ros_platformio.h>
+
+static uint16_t rhead, rtail;
+static volatile uint16_t thead, thead_next, ttail;
+static uint32_t open_count = 0;
+
+static void stream_flush(DMAStream *stream) {
+  static volatile bool mutex = false;
+
+  if ((stream->uart->gState == HAL_UART_STATE_READY) && !mutex) {
+    mutex = true;
+
+    if (thead != ttail) {
+      uint16_t len =
+          thead < ttail ? ttail - thead : stream->tbuffer_size - thead;
+      thead_next = (thead + len) % stream->tbuffer_size;
+      HAL_UART_Transmit_DMA(stream->uart, stream->tbuffer + thead, len);
+    }
+    mutex = false;
+  }
+}
+
+void uart_transfer_complete_callback(DMAStream *stream) {
+  thead = thead_next;
+  stream_flush(stream);
+}
+
+extern "C" {
+
+bool platformio_transport_open(struct uxrCustomTransport *transport) {
+  DMAStream *stream = (DMAStream *)transport->args;
+
+  if (open_count == 0) {
+    rhead = rtail = thead = thead_next = ttail = 0;
+    HAL_UART_Receive_DMA(stream->uart, stream->rbuffer, stream->rbuffer_size);
+  }
+  open_count++;
```

Comment:
Why use a integer for this?

Also, did you encounter any situation where open/close calls miss match?

Reply:
In my case, I used `rmw_uros_ping_agent` periodically to detect agent disconnects. The issue was that this function called open and close on the transport each time it was called, even when the transport was already opened. This is why I decided to ignore all nested open/close calls, hence I use integer to track the nesting level.

It's possible I used the library wrong. Let me know if that's the case or if you know a better way to deal with this problem.